contracts/upgradeable_contracts/ERC20ExtendedBridge.sol

```diff
@@ -0,0 +1,19 @@
+pragma solidity 0.4.24;
+
+
+import "./BasicBridge.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20Basic.sol";
```

Comment:
We have already very a similar interface `ERC677Bridge`. But in the new logic it makes sense to call it `ERC677BridgeForBurnableMintableToken`, doesn't it?

Below is my suggestion how to re-organize the hierarchy of the contracts in this case.

*ERC677Bridge.sol*
```solidity
pragma solidity 0.4.24;


import "./BasicBridge.sol";
import "../ERC677.sol";

contract ERC677Bridge is BasicBridge {
    function erc677token() public view returns(ERC677) {
        return ERC677(addressStorage[keccak256(abi.encodePacked("erc677token"))]);
    }

    function setErc677token(address _token) internal {
        require(_token != address(0) && isContract(_token));
        addressStorage[keccak256(abi.encodePacked("erc677token"))] = _token;
    }

    function onTokenTransfer(address _from, uint256 _value, bytes /*_data*/) external returns(bool) {
    	ERC677 token = erc677token();
        require(msg.sender == address(token));
        require(withinLimit(_value));
        setTotalSpentPerDay(getCurrentDay(), totalSpentPerDay(getCurrentDay()).add(_value));
        bridgeSpecificActionsOnTokenTransfer(token, _from, _value);
        return true;
    }

    function bridgeSpecificActionsOnTokenTransfer(ERC677 _token, address _from, uint256 _value) internal {
        fireEventOnTokenTransfer(_from, _value);
    }

    function fireEventOnTokenTransfer(address /*_from */, uint256 /* _value */) internal;

}
```

*ERC677BridgeForBurnableMintableToken.sol*
```solidity
pragma solidity 0.4.24;


import "./BasicBridge.sol";
import "./ERC677Bridge.sol";

contract ERC677BridgeForBurnableMintableToken is ERC677Bridge {
    function bridgeSpecificActionsOnTokenTransfer(IBurnableMintableERC677Token _token, address _from, uint256 _value) internal {
        erc677token().burn(_value);
        fireEventOnTokenTransfer(_from, _value);
    }

    function fireEventOnTokenTransfer(address /*_from */, uint256 /* _value */) internal;

}
```

* *erc20_to_erc20/HomeBridgeErcToErc.sol* will inherit *ERC677BridgeForBurnableMintableToken* rather than *ERC677Bridge*
* *native_to_erc20/ForeignBridgeNativeToErc.sol* will inherit *ERC677BridgeForBurnableMintableToken* rather than *ERC677Bridge*
* *erc20_to_erc20/ForeignBridgeErc677ToErc677.sol*
```solidity
pragma solidity 0.4.24;

import "./ForeignBridgeErcToErc.sol";
import "../ERC677Bridge.sol";

contract ForeignBridgeErc677ToErc677 is ERC677Bridge, ForeignBridgeErcToErc {

    event UserRequestForAffirmation(address recipient, uint256 value);

    function initialize(
        address _validatorContract,
        address _erc20token,
        uint256 _requiredBlockConfirmations,
        uint256 _gasPrice,
        uint256 _dailyLimit,
        uint256 _maxPerTx,
        uint256 _minPerTx,
        uint256 _homeDailyLimit,
        uint256 _homeMaxPerTx,
        address _owner
    ) public returns(bool) {
        require(_minPerTx > 0 && _maxPerTx > _minPerTx && _dailyLimit > _maxPerTx);

        _initialize(
            _validatorContract,
            _erc20token,
            _requiredBlockConfirmations,
            _gasPrice,
            _homeDailyLimit,
            _homeMaxPerTx,
            _owner
        );

        uintStorage[keccak256(abi.encodePacked("dailyLimit"))] = _dailyLimit;
        uintStorage[keccak256(abi.encodePacked("maxPerTx"))] = _maxPerTx;
        uintStorage[keccak256(abi.encodePacked("minPerTx"))] = _minPerTx;

        return isInitialized();
    }

    function erc20token() public view returns(ERC20Basic) {
        return erc677token();
    }

    function setErc20token(address _token) private {
        setErc677token(_token);
    }

    function fireEventOnTokenTransfer(address _from, uint256 _value) internal {
        emit UserRequestForAffirmation(_from, _value);
    }
}
```

Reply:
Thanks for the detailed explanation! Here I applied the changes https://github.com/poanetwork/poa-bridge-contracts/pull/177/commits/5d54ffb20bda80cf64a59529569bec4f1ef60ba5